Fixed issue #15940: Export as printable : PHP Notice wuth debug

gabrieljenik opened the pull request:
When using global template on a survey, the template name saved on DB is "inherit".
That was taken literal sometimes and caused errors, as there is no template called "inherit"

"inherit" template is now converted to the proper template name, set at global level.

Shnoulle commented:
Question : what happen if you name a template `inherit` 🤔 another issue to report ?

Shnoulle commented:
In  my opinion : there are an issue before … (but this one is a fix) $oSurvey-template MUST never be set to an inexisting survey. `getTemplatePath` and `getTemplateURL` must not be called with an invalid template :)

gabrieljenik commented:
> Question : what happen if you name a template `inherit`  another issue to report ?

Yes, noted that on the bug

gabrieljenik commented:
> In my opinion : there are an issue before … (but this one is a fix) $oSurvey-template MUST never be set to an inexisting survey. `getTemplatePath` and `getTemplateURL` must not be called with an invalid template :)

Will fix that. We should throw an exception, right?

Shnoulle commented:
> > In my opinion : there are an issue before … (but this one is a fix) $oSurvey-template MUST never be set to an inexisting survey. `getTemplatePath` and `getTemplateURL` must not be called with an invalid template :)
> 
> Will fix that. We should throw an exception, right?

If you find the other issue, but don't take to more time on it :)

Yes : think best is to throw clean error if template is invalid. But it's **my** opinion.

I think too : if we get there and template is inavlid : maybe some css/js was not loaded ( theme package)

gabrieljenik commented:
> If you find the other issue, but don't take to more time on it :)

Here you mean?
https://github.com/LimeSurvey/LimeSurvey/pull/1449/files#diff-66fb738711f2db21da8c5b6029ad0794R295
If oTemaplte is empty, throw error.

Error message could be "Survey template not found"?

Shnoulle commented:
> > If you find the other issue, but don't take to more time on it :)
> 
> Here you mean?
> https://github.com/LimeSurvey/LimeSurvey/pull/1449/files#diff-66fb738711f2db21da8c5b6029ad0794R295
> If oTemaplte is empty, throw error.
> 
> Error message could be "Survey template not found"?

Not exactly : i mean : template must be checked before this function. Then : why the template is not OK for print version.

gabrieljenik commented:
> Not exactly : i mean : template must be checked before this function. Then : why the template is not OK for print version.

Oh, I agree. Template must be checked. Actually the template name comes out from the DB. So it is kind of checked. Also templateNameFilter() does lot of checkings. Still, I suggest the "Survey template not found" exception as a sanity check (I think they are called like that), just in case. Don't expect to be thrown, but better to ave an exception than a dump.

Where did you picture an extra validation?

Shnoulle commented:
> Actually the template name comes out from the DB.

in Survey->find ? Then it must be already fixed … else all is broken.

> Don't expect to be thrown, but better to ave an exception than a dump.

Throw an exception id really the best solution :)

> Where did you picture an extra validation?

I don't want extra validation : only Survey->find must validate it 

Currently : inherit is disable of fixing … 

https://github.com/LimeSurvey/LimeSurvey/blob/18f3377f2e0eb94ba4bc6ee29438bc8b4d5ba0df/application/models/Survey.php#L565-L567

And this in an issue for Print …

gabrieljenik commented:
> And this in an issue for Print …

Agree this is what is making things break.
Maybe we can have a method on Survey model, like "GetEffectiveTemplate()" or "GetEffectiveTemplateName()" ?

Shnoulle commented:
I don't know how it done with Survey, but maybe this explan why sometimes it broke at 1st call, but not at second call (when delete or import some specific survey)

gabrieljenik commented:
>Maybe we can have a method on Survey model, like "GetEffectiveTemplate()" or "GetEffectiveTemplateName()" ?

As to not loose track fo the conversation, @olleharstedt @Shnoulle Should I do this?

olleharstedt commented:
What happens if I create a theme with name "inherit"?

olleharstedt commented:
What's wrong with this fix, exactly?

gabrieljenik commented:
> What's wrong with this fix, exactly?

Nothing is wrong really. Fix should work good.
I was saying that:
- If the survey uses the global default theme, "inherit" is saved on the DB
- If that's the case, the survey model doesn't validate the theme when laoding a survey record.
- When using the global theme, as to be able to get the effective theme to be used, I was thinking about adding a method.
But I realize think it is out of the scope...

gabrieljenik commented:
> > If you find the other issue, but don't take to more time on it :)
> 
> Here you mean?
> https://github.com/LimeSurvey/LimeSurvey/pull/1449/files#diff-66fb738711f2db21da8c5b6029ad0794R295
> If oTemaplte is empty, throw error.
> 
> Error message could be "Survey template not found"?

Done

Shnoulle commented:
Still an isue in this fix : if template is set as inherit (from Survey group) : it's default template used.

I think this commit didn't fix the issue

Shnoulle commented:
Survey use `$oTemplate = Template::model()->getInstance('', $surveyid);`

printable version **MUST** use `$oTemplate = Template::model()->getInstance('', $surveyid);`

Shnoulle commented:
Your commit didn't take inherited template , only default template

https://github.com/LimeSurvey/LimeSurvey/pull/1455

olleharstedt commented on application/models/Template.php:
```diff
@@ -293,6 +293,10 @@ public static function getTemplatePath($sTemplateName = "")
         }
 
         $oTemplate = self::model()->findByPk($sTemplateName);
+        if (empty($oTemplate))
```
https://www.php-fig.org/psr/psr-12/

Should be

```php
if (empty($oTemplate)) {
```

`{` on the same line. You should include PSR-12 checking in your editor (will show LOTS of errors, but make sure all new code follows it, please).

gabrieljenik replied:
Fixed

olleharstedt commented on application/models/Template.php:
```diff
@@ -293,6 +293,10 @@ public static function getTemplatePath($sTemplateName = "")
         $oTemplate = self::model()->findByPk($sTemplateName);
+        if (empty($oTemplate))
+        {
+            throw new \Exception("Survey theme {$sTemplateName} not found.", 1);
```
Not required, but if you want to avoid `\`, you can add `use Exception;` at top of file (if there's a namespace, not sure there is right now).

olleharstedt replied:
Something to discuss is which exception messages should be translated with gT().

Shnoulle replied:
Must use CException since we use Yii (and there are no namespace)

olleharstedt replied:
OK, sure.

gabrieljenik replied:
Fixed

Shnoulle commented:
@olleharstedt : did you read my comment ? If we want to use template we NEED to get inherited template. Else : it's another issue … 

https://github.com/LimeSurvey/LimeSurvey/pull/1455

gabrieljenik commented:
Added method/attribute on survey model for getting the effective template name (handling survey group inheritance). Used that method/attribute while exporting printable surveys.

Added some existence checkings on getTemplateUrl() and getTemplatePath()

gabrieljenik commented:
> @olleharstedt : did you read my comment ? If we want to use template we NEED to get inherited template. Else : it's another issue …
> 
> #1455

Please review my latest commit. I think it goes in the same direction of your PR

Shnoulle commented on application/models/Survey.php:
```diff
@@ -835,6 +835,33 @@ public function getSurveyTemplateConfiguration()
+            if (!empty($this->oOptions->template)) {
+                $sTemplateName = $this->oOptions->template;
+            } else {
+                throw new CException("Unable to get a template name from group for survey {$this->sid}");
```
I like the new function :+1: .

But here : i don't know how it work for parent : parent can not be inherit too ? Then get it for "global" ?

gabrieljenik replied:
I actually tooke the code from digging into your example.
The code is from here: https://github.com/gabrieljenik/LimeSurvey/blob/65061857c0ae8514300746b481285e0004a7d576/application/models/TemplateConfiguration.php#L259

So, I think survey group inheritance is there. 
Now that you mention it, not sure how it would work with multiple survey groups inherted.
Is that possible? Does the TemplateConfiguration code supports it?

@Shnoulle What I don't like that much your intial approach (sorry :) ) is that there is a lot of processing (template instancing) just for getting the template name.

Shnoulle replied:
Thanks : never look at getInstanceFromSurveyId :)

olleharstedt commented:
Looks good to me! Denis, you happy?

Shnoulle commented:
I don't really know what happen fro group , i don't know if there are a way where group can return inherit.
I don't work on inherit system : then don't really know if we can have a more global function.

But think it can be checked after.

sushmanadendlalimesurvey commented:
Tested by pulling the PR. Able to successfully download and view the survey in debug mode

olleharstedt commented:
Gah, now I merged with the typo D:

gabrieljenik commented:
> I don't really know what happen fro group , i don't know if there are a way where group can return inherit.
> I don't work on inherit system : then don't really know if we can have a more global function.
> 
> But think it can be checked after.

Agree: We can make the new method thicker if needed after.

Shnoulle commented:
> > I don't really know what happen fro group , i don't know if there are a way where group can return inherit.
> > I don't work on inherit system : then don't really know if we can have a more global function.
> > But think it can be checked after.
> 
> Agree: We can make the new method thicker if needed after.

And since : you create a clean function : really more easy to fix :)

gabrieljenik commented:
> And since : you create a clean function : really more easy to fix :)

Thanks for the comment :)
